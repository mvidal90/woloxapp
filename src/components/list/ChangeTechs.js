import React, { useState } from 'react';

export const ChangeTechs = ({ setTechs }) => {

    const [inputValue, setInputValue] = useState('hola');

    const handleSubmit = (e) => {
        e.preventDefault();

        setTechs( techs => [ inputValue ]);
    }

    const handleInputChange = ( e ) => {
        setInputValue(e.target.value);
    }

    return (
        <div className = "container">
            <div className = "row">
                <form onSubmit={ handleSubmit }>
                    <input 
                        type = "Text"
                        placeholder = "Busca tu tecnologÃ­a"
                        value = { inputValue }
                        onChange = { handleInputChange }
                    />
                </form>
            </div>
        </div>
    )
}
