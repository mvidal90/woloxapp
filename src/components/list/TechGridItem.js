import React, { useEffect, useState } from 'react';

export const TechGridItem = ({
    tech,
    year,
    author,
    license,
    language,
    type,
    logo,
    stg,
    favorites,
    setFavorites,
    handleFavorites
}) => {

    const [isFavorite, setIsFavorite] = useState(false);

    const add = () =>{
        setIsFavorite( true );
        if (favorites === null) {
            setFavorites([tech]);
        } else {
            setFavorites([...favorites, tech]);
        }
    };

    const del = () =>{
        setIsFavorite( false );
        setFavorites(favorites.filter(e => e !== tech))
    };

    useEffect(() => {

        if (stg !== null) {
            if (stg.includes( tech )) {
            
                setIsFavorite( true );
            }
        }
    }, [stg, tech])
    
    
    return (
        <div  className="card">
            <div className = "logos-container">
                <img src={ logo } alt={ tech } className = "img-size" />
            </div>
            <div className = "data-container">
                <h4>{ tech } ( { year } )</h4>
                <p><strong>Autor: </strong> { author } </p> 
                <p><strong>Licencia: </strong> { license } </p> 
                <p><strong>Lenguaje: </strong> { language } </p> 
                <p><strong>Tipo: </strong> { type } </p> 
            </div>

            <form onSubmit = { handleFavorites }>
                {
                    isFavorite ?
                        <button className = "btn-tch2" onClick = { del }>Quitar de Favoritos</button>
                    :
                        <button className = "btn-tch" onClick = { add } >Agregar a Favoritos</button>
                }
            </form>

        </div>
    )
}
