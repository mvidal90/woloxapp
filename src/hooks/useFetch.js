import { useState, useEffect, useRef } from 'react';

export const useFetch = ( url ) => {

    const isMonted = useRef(true);
    const [state, setState] = useState({ data: null, loading: true, error: null});

    useEffect(() => {
        return() =>{
            isMonted.current = false;
        }
    }, [])

    useEffect(() => {
        
        setState({ data:null, loading:true, error:null});

        fetch( url )
            .then( resp => resp.json() )
            .then( data=>{

                if ( isMonted.current){
                    setState({
                        loading: false,
                        error: null,
                        data
                    });
                }
            })
            .catch( () => {
                setState({
                    data: null, 
                    loading: false, 
                    error: 'No se pudo cargar la info'
                })
            })

    }, [url]);

    return state;
}
