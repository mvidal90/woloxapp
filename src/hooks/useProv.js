import { useState, useEffect } from "react"
import { getProv } from '../helpers/getProv';

export const useProv = ( country ) =>{

    const [stateProvs, setStateProvs] = useState({
        provincies: [],
    });

    useEffect( () =>{
                         
        const { provincies } = getProv( country );
        setStateProvs(provincies);

    }, [country])

    return stateProvs;

}