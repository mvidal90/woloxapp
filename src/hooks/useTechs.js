import { useState, useEffect } from "react"

export const useTechs = () =>{

    const [state, setState] = useState({
        data: [],
        loading: true
    });

    useEffect( () => {
        
        var request = new XMLHttpRequest();
    
        request.open('GET', 'http://private-8e8921-woloxfrontendinverview.apiary-mock.com/techs?tech=Node');
    
        request.onreadystatechange = function( e ) {
            if (this.readyState === 4) {

                setState({
                    data: JSON.parse(this.responseText),
                    loading: false
                })      
    
            }
        };
        
        request.send();
            
    }, [])

    return state;

}