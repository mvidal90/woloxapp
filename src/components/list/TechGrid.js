import React, { useState } from 'react';
import { useTechs } from '../../hooks/useTechs';
import { TechGridItem } from './TechGridItem';

export const TechGrid = ( {stg} ) => {

    const { data, loading } = useTechs();
    const [favorites, setFavorites] = useState(JSON.parse(localStorage.getItem('favorites')))

    const handleFavorites = (e) => {

        localStorage.setItem('favorites', JSON.stringify( favorites ))

    }

    return (    
    <>
        { loading && <p>Loading</p> }
        <div className = "wrapper">
        
            {
                data.map( tech => (
                    <TechGridItem 
                        stg = { stg }
                        favorites = { favorites }
                        setFavorites = { setFavorites }
                        handleFavorites = { handleFavorites }
                        key={ tech.tech }
                        { ...tech }
                    />
                )) 
            }
        
        </div>
        <div>
            <p>Total de tecnolog√≠as: {data.length}</p>
        </div>
    </>       
)
}
