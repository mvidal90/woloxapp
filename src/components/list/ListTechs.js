import React, { useEffect } from 'react';
import { ChangeTechs } from './ChangeTechs';
import { TechGrid } from './TechGrid';

export const ListTechs = ({ setHome, setMsg }) => {

    let stg = JSON.parse(localStorage.getItem('favorites'));

    useEffect(() => {
        setHome( false );
        if (!localStorage.getItem('favorites')) {
            setMsg(`Favoritos: 0`);
        } else {
            setMsg(`Favoritos: ${ stg.length }`);
        }
    }, [stg, setHome, setMsg]);

    return (
        <section id ="Techs">
            <div className = "container">
                <div className = "row">
                    <h2> WoloxApp </h2>
                    {/* <ChangeTechs setTechs = { setTechs }/> */}
                    <ol>
                        <TechGrid  stg = { stg }/>
                    </ol>
                </div>
            </div>
        </section>
    );
}
