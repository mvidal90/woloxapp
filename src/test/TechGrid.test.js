import React from 'react';
import { shallow } from 'enzyme';
import { TechGrid } from '../components/list/TechGrid';

import { useTechs } from '../hooks/useTechs';
jest.mock('../hooks/useTechs');

describe('Pruebas con <TechGrid />', () => {

    test('Debe mostrarse correctamente mientras se consiguen los ítems', () => {

        useTechs.mockReturnValue({
            data:[],
            loading: true
        });
          
        const wrapper = shallow( <TechGrid /> );

        expect( wrapper ).toMatchSnapshot(); 
        expect( wrapper.find('TechGridItem').length).toBe(0);

    })

    test('Debe mostrarse correctamente cuando se muestran los ítems', () => {

        useTechs.mockReturnValue({
            data:[
                {
                  tech: "Node",
                  year: "2009",
                  author: "Ryan Dahl",
                  license: "MIT",
                  language: "JavaScript",
                  type: "Back-End",
                  logo: "https://upload.wikimedia.org/wikipedia/commons/thumb/d/d9/Node.js_logo.svg/220px-Node.js_logo.svg.png"
                },
                {
                  tech: "React",
                  year: "2013",
                  author: "Jordan Walke",
                  license: "MIT",
                  language: "JavaScript",
                  type: "Front-End",
                  logo: "https://upload.wikimedia.org/wikipedia/commons/thumb/a/a7/React-icon.svg/220px-React-icon.svg.png"
                },
                {
                  tech: "Vue",
                  year: "2014",
                  author: "Evan You",
                  license: "MIT",
                  language: "JavaScript",
                  type: "Front-End",
                  logo: "https://upload.wikimedia.org/wikipedia/commons/thumb/9/95/Vue.js_Logo_2.svg/220px-Vue.js_Logo_2.svg.png"
                }],
            loading: false
        });
          
        const wrapper = shallow( <TechGrid /> );

        expect( wrapper ).toMatchSnapshot(); 
        expect( wrapper.find('TechGridItem').length).toBe(3);

    })

})
