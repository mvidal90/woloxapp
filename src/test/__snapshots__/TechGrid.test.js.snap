// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas con <TechGrid /> Debe mostrarse correctamente cuando se muestran los ítems 1`] = `
<Fragment>
  <div
    className="wrapper"
  >
    <TechGridItem
      author="Ryan Dahl"
      favorites={null}
      handleFavorites={[Function]}
      key="Node"
      language="JavaScript"
      license="MIT"
      logo="https://upload.wikimedia.org/wikipedia/commons/thumb/d/d9/Node.js_logo.svg/220px-Node.js_logo.svg.png"
      setFavorites={[Function]}
      tech="Node"
      type="Back-End"
      year="2009"
    />
    <TechGridItem
      author="Jordan Walke"
      favorites={null}
      handleFavorites={[Function]}
      key="React"
      language="JavaScript"
      license="MIT"
      logo="https://upload.wikimedia.org/wikipedia/commons/thumb/a/a7/React-icon.svg/220px-React-icon.svg.png"
      setFavorites={[Function]}
      tech="React"
      type="Front-End"
      year="2013"
    />
    <TechGridItem
      author="Evan You"
      favorites={null}
      handleFavorites={[Function]}
      key="Vue"
      language="JavaScript"
      license="MIT"
      logo="https://upload.wikimedia.org/wikipedia/commons/thumb/9/95/Vue.js_Logo_2.svg/220px-Vue.js_Logo_2.svg.png"
      setFavorites={[Function]}
      tech="Vue"
      type="Front-End"
      year="2014"
    />
  </div>
  <div>
    <p>
      Total de tecnologías: 
      3
    </p>
  </div>
</Fragment>
`;

exports[`Pruebas con <TechGrid /> Debe mostrarse correctamente mientras se consiguen los ítems 1`] = `
<Fragment>
  <p>
    Loading
  </p>
  <div
    className="wrapper"
  />
  <div>
    <p>
      Total de tecnologías: 
      0
    </p>
  </div>
</Fragment>
`;
